feat: add How to override or customize native section components (#16)
vtex/faststore-docs: docs/tutorials/how-to-override-section-components.md

---
title: "How to override or customize native section components"
sidebar_label: "How to override or customize native section components"
---

# How to override or customize native section components

> This guide covers how to override native section components or pass additional props to them.

Reviewer: ```suggestion
In this guide, learn two types of customization of a native section component: overriding and passing additional props to them.
```

> Native sections are provided by default by VTEX's `@faststore/core` package. Sections are composed by different components that can be customized or overriden to add functionalities or to change looks.

Reviewer 2: Add also a fake link to where our list of sections will be.

> [Native sections](/tbd-link-to-native-sections-list) are provided by default by VTEX's `@faststore/core` package. Sections are composed by different components that can be customized or overriden to add functionalities or to change looks.

Reviewer: ```suggestion
[Native sections](/tbd-link-to-native-sections-list) are composed of different components that can be customized or overridden to add functionalities or to change their appearance in the storefront. Sections are provided by default by VTEX's `@faststore/core` package.
```

{Image or gif of the expected behavior after completing the guide}

---

## Before you start

_Before you begin, you should already have:_
- A functioning Evergreen store.

_You should also consider that:_

> - This feature is experimental.

Reviewer 2: This won't be experimental when we launch, will it?

Reviewer 3: I think it's ok to keep it like that. It's safer to say that it is an experimental feature 👍 hihi

Reviewer: Since it's experimental, does it require any special config. or any permission? Can we list any behavior the dev could expect working with this feature?

- Some sections contain multiple instances of the same component. When overriting a component or passing additional props to it, consider that all instances will be affected by this.

---

## Step by step
### Step 1 - Choose a native section to customize

1. Choose the native section to be customized - [list of available native sections](./tbd-link)

> 2. Navigate to `src/components/overrides` and create a new file named after the native section

Reviewer: ```suggestion
2. Navigate to `src/components/overrides` and create a new file named after the native section. For example, if you chose the `ProductDetails` section for customization, create a new file named `ProductDetails.tsx` under the `src/components/overrides` directory.
```

> eg. Consider you chose the `ProductDetails` section to customize. Create a new file named `ProductDetails.tsx` under the `src/components/overrides` directory.

Reviewer 2: ```suggestion
eg. Considering you chose the `ProductDetails` section to customize. Create a new file named `ProductDetails.tsx` under the `src/components/overrides` directory.
```

Author: I think it sits better as "Consider", since this is an arbitrary choice and just an example. "Considering" makes me think the reader had a set of options (it has, but it's not presented in the text) and that the possibility of them choosing other alternatives will be addressed later. It also reads as a suggestion, which is accurate in this scenario.

> 3. Copy and paste the following boilerplate on the file

Reviewer: ```suggestion
3. Copy and paste the following boilerplate on the file: 
```

You should change the value of the `SECTION` variable to the name of the section you chose to override. In this case, `ProductDetails` is used.

```tsx
const SECTION = "ProductDetails";

const overrides: SectionOverride[SECTION] = { name: SECTION, components: {} };

export default overrides;
```

> 

Reviewer: ```suggestion
> ⚠️ Change the value of the `SECTION` variable to the name of the section you want to override. In this case, we used the `ProductDetails` example.
```

### Step 2 - Choose a component to override

> 1. Choose a component to override

Reviewer: ```suggestion
1. Choose a component to override from the [list of overridable components for each native section](tbd-link). In this example, we are overriding the `Price` component.
```

You can see a list of overridable components for each native section [here](tbd-link). For this guide, we're choosing to override the `Price` component.

> 2. Add it to the `overrides` object

Reviewer: ```suggestion
2. Add an object with the name of the component you wish to override to the `components` property inside the `overrides` object.
```

Add an object with the name of the component you wish to override to the `components` property inside the `overrides` object.

> For this guide, we're choosing to override the `Price` component.

Reviewer: > 1. Choose a component to override from the [list of overridable components for each native section](tbd-link). In this example, we are overriding the `Price` component.

If we already stated at the beginning of this step that we are using the `Price` component, we don't need to repeat it.

```tsx
const SECTION = "ProductDetails"

const overrides: SectionOverride[SECTION] = {
  name: SECTION,
  components: {
    Price: {},
  },
}

export default overrides
```

> ### Step 3 - Override the component or pass additional props to a it

Reviewer: ```suggestion
### Step 3 - Override or pass additional props to a component
```

> 1. Choose if you want override the component or pass additional props to it

Reviewer: ```suggestion
Overriding a native component may be necessary if it lacks the functionality you require or has a layout that does not meet your needs. There are cases where a prop is available to customize the behavior of the native component. Additionally, you can customize your store's theme to tailor the look and feel of the native sections to your needs. 

For more information on when to override or pass additional prop to a component, and the best practices for overriding components, check the following documentation:

- [doc 1]()
- [doc 2]()

> ⚠️ Overriding a component leads you to miss out on performance updates and bug fixes provided by VTEX for that component. By passing an additional prop and using our theming tools, you can achieve the desired behavior without missing out on updates and bug fixes from VTEX. You can find the available props for each component [here](https://evergreen.faststore.dev/components).
```

There's a great documentation explaining when you want to override the component or just pass additional props to it - here's the [link](tbd-link).

In summary, you want to override the native component when it doesn't offer the functionality you need or when it is layed out much different from what you desire. By doing that, you'll be losing on the performance updates and bug fixes provided by VTEX for that component. For more in-depth information regarding best practices when overriding components, check the link [here](tbd-link).

> For most common use cases, there's probably a prop that can be used to customize the behavior of the native component. By passing an additional prop, you get the behavior you originally wanted and don't miss out on updates and bug fixes released by VTEX. You can see the list of available props for each component [here](/tbd-link).

Reviewer 2: Mention instead of prop but the combo of prop + themification can give you control over look and feel + behavior

Reviewer 3: maybe you can link https://evergreen.faststore.dev/components for `available props` for each component

2. (Optional) Override the component

> Inside the `Price` object, add a property called `Component`. This property value should be defined as a React component created by you, that'll be used to replace every instance of the `Price` component inside the `ProductDetails` section.

Reviewer: ```suggestion
Inside the `Price` object, add a property called `Component`. This property value should be defined as a React component created by you, that will be used to replace every instance of the `Price` component inside the `ProductDetails` section.
```

```tsx
import CustomPrice from "src/components/custom/CustomPrice"

const SECTION = "ProductDetails"

const overrides: SectionOverride[SECTION] = {
  name: SECTION,
  components: {
    Price: {
      Component: CustomPrice,
    },
  },
}

export default overrides
```

3. (Optional) Pass additional props to the component

Inside the `Price` object, add a property called `props`. In addition to the default props this component receives from the section it is contained, every property defined inside the props object will also be passed to it. If a prop you passed has the same name as an already existing prop being received by the component, yours will override the default definition.

```tsx
const SECTION = "ProductDetails"

const overrides: SectionOverride[SECTION] = {
  name: SECTION,
  components: {
    Price: {
      props: {
        variation: "listing",
      },
    },
  },
}

export default overrides
```

## Summary

You can have different outcomes depending if you chose to override the component or pass additional props to it.

> > If you override the component, additional props will not be considered when overriding it.

Reviewer: ```suggestion
> ⚠️ If you override the component, additional props will not be considered when overriding it.
```
